internal/provider/services/storage/storage.go

```diff
@@ -0,0 +1,235 @@
+		attribute.String("provider.version", request.GetProvider().GetVersion()),
+		attribute.String("provider.os", request.GetProvider().GetOs()),
+		attribute.String("provider.arch", request.GetProvider().GetArch()),
+	)
+
+	providerAddress := strings.Split(request.Provider.GetName(), "/")
+	filename := fmt.Sprintf("terraform-provider-%s_%s_%s_%s.zip", providerAddress[1], request.GetProvider().GetVersion(), request.GetProvider().GetOs(), request.GetProvider().GetArch())
+	fileLocation := services.ResolveS3Locations(request.Provider.GetName(), request.Provider.GetVersion(), filename)
+
+	in := &s3.GetObjectInput{
+		Bucket: aws.String(BucketName),
+		Key:    aws.String(fileLocation),
+	}
+
+	out, err := s.Client.GetObject(ctx, in)
+	if err != nil {
+		span.RecordError(err)
+		log.Println(err)
+		return DownloadSourceZipError
+	}
+
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(out.Body)
+	bb := buf.Bytes()
+
+	outContentLength := out.ContentLength
+
+	if int64(len(bb)) == outContentLength {
+		res := &services.SourceZipResponse{}
+		for i := int64(0); i < outContentLength; i += ChunkSize {
+			if i+ChunkSize > outContentLength {
+				res.ZipDataChunk = bb[i:outContentLength]
+			} else {
+				res.ZipDataChunk = bb[i : i+ChunkSize]
+			}
+
+			if err := server.Send(res); err != nil {
+				span.RecordError(err)
+				log.Println(err)
+				return SendSourceZipError
+			}
+		}
+
+		log.Println("Source zip downloaded.")
+		return nil
+	} else {
+		span.RecordError(err)
+		return ContentLengthError
+	}
+}
+
+// Download Shasum from storage
+func (s *StorageService) DownloadShasum(request *services.DownloadShasumRequest, server services.Storage_DownloadShasumServer) error {
+
+	log.Println("Downloading shasum file.")
+
+	ctx := server.Context()
+	span := trace.SpanFromContext(ctx)
+	span.SetAttributes(
+		attribute.String("provider.name", request.GetProvider().GetName()),
+		attribute.String("provider.version", request.GetProvider().GetVersion()),
+	)
+
+	providerAddress := strings.Split(request.GetProvider().GetName(), "/")
+	prefix := fmt.Sprintf("terraform-provider-%s_%s_SHA256SUMS", providerAddress[1], request.GetProvider().GetVersion())
+	fileLocation := services.ResolveS3Locations(request.Provider.GetName(), request.Provider.GetVersion(), prefix)
+
+	in := &s3.GetObjectInput{
+		Bucket: aws.String(BucketName),
+		Key:    aws.String(fileLocation),
+	}
+
+	out, err := s.Client.GetObject(ctx, in)
+	if err != nil {
+		span.RecordError(err)
+		log.Println(err)
+		return DownloadShasumError
+	}
+
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(out.Body)
+	bb := buf.Bytes()
+
+	outContentLength := out.ContentLength
+
+	if int64(len(bb)) == outContentLength {
+		res := &services.DownloadShasumResponse{}
+		for i := int64(0); i < outContentLength; i += ChunkSize {
+			if i+ChunkSize > outContentLength {
+				res.ShasumDataChunk = bb[i:outContentLength]
+			} else {
+				res.ShasumDataChunk = bb[i : i+ChunkSize]
+			}
+
+			if err := server.Send(res); err != nil {
```

Comment:
Same comment here with regard to reading the whole file and then chunking it.
Simply read the data chunk at a time then send that chunk and repeat.

Reply:
updated